src/api/trace/span_context.rs

```diff
@@ -111,6 +114,110 @@ impl SpanId {
     }
 }
 
+/// TraceState carries system-specific configuration data, represented as a list
+/// of key-value pairs. TraceState allows multiple tracing systems to
+/// participate in the same trace.
+///
+/// Please review the [W3C specification] for details on this field.
+///
+/// [W3C specification]: https://www.w3.org/TR/trace-context/#tracestate-header
+#[cfg_attr(feature = "serialize", derive(Deserialize, Serialize))]
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct TraceState {
+    data: HashMap<String, String>,
+    ordered_keys: VecDeque<String>,
+}
+
+impl TraceState {
+    /// Creates a new `TraceState` from the given key-value collection.
+    pub fn from_key_value(trace_state: Vec<KeyValue>) -> Self {
+        let mut data: HashMap<String, String> = HashMap::with_capacity(trace_state.len());
```

Comment:
If you were to go with `BTreeMap` this could be simplified to something like:

```rust
pub fn new<T: IntoIterator<Item = (String, String)>>(trace_state: T) -> Self {
    TraceState {
        data: trace_state.into_iter().collect(),
    }
}
```